### PR TITLE
[Merged by Bors] - feat(Analysis/Analytic): A few lemmas that simple things are analytic

### PR DESCRIPTION
**Pull request by girving**
We record various simple lemmas that things are `analyticAt` or `analyticOn`.  There's no hard work here, just corollaries of other results:

1. `id`, `fst`, `snd`
2. Power series terms, in the origin (we already know they have power series, this is just the `.analyticAt_changeOrigin` corollary
3. Finite sums
4. Pairs of analytic functions: `x ↦ (f x, g x)`

We also add a few lemmas for dealing with curried analytic functions. Starting with `AnalyticOn 𝕜 (uncurry h) s`,

1. `AnalyticOn.curry_comp` composes it with two input analytic functions
2. `AnalyticOn.along_fst/snd` show analyticity along each coordinate

---

This is the first (extremely minimal) PR for upstreaming of https://github.com/girving/ray: Sebastien Gouezel suggested lemmas about analyticity, so I started with the simplest ones.  The next thing I would upstream after this is `AnalyticAt.mul`, but that requires some `ContinuousMultilinearMap` machinery so I'll do it as a second PR unless advised otherwise.

### CODE REVIEW

**Comment by ADedecker** [comment]
Thanks for your first PR! Is it ready for review yet? If so, please add the "awaiting-review" and "t-analysis" labels, otherwise this PR won't appear in the list of PRs to review that most maintainers use :wink: See https://leanprover-community.github.io/contribute/index.html#lifecycle-of-a-pr for more details.

File: Mathlib/Analysis/Analytic/Linear.lean
Comment (semorrison) [inline Mathlib/Analysis/Analytic/Linear.lean:171]:
```suggestion
/-- `fst` is entire -/
```

File: Mathlib/Analysis/Analytic/Linear.lean
Comment (semorrison) [inline Mathlib/Analysis/Analytic/Linear.lean:175]:
```suggestion
/-- `snd` is entire -/
```

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (semorrison) [inline Mathlib/Analysis/Analytic/Prod.lean:55]:
```suggestion
/-- Products of analytic functions are analytic -/
```

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (semorrison) [inline Mathlib/Analysis/Analytic/Prod.lean:73]:
```suggestion
    (m : ∀ x, x ∈ t → (f x, g x) ∈ s) : AnalyticOn 𝕜 (fun x ↦ h (f x) (g x)) t := 
  fun _ xt ↦ (ha _ (m _ xt)).curry_comp (fa _ xt) (ga _ xt)
```
I think we tend not to keep the `fun` on the `:=` line, although probably Mathlib is inconsistent about this.

File: Mathlib/Analysis/Analytic/Basic.lean
Comment (semorrison) [inline Mathlib/Analysis/Analytic/Basic.lean:1477]: Not completely essential, but I think it's good style to use `· ` every time there are multiple goals.

File: Mathlib/Analysis/Analytic/Basic.lean
Comment (eric-wieser) [inline Mathlib/Analysis/Analytic/Basic.lean:1470]: Does this generalize to a Finset of an arbitrary type?

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Prod.lean:73]: Is there a way to lint for this?

File: Mathlib/Analysis/Analytic/Basic.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Basic.lean:1470]: Done.

File: Mathlib/Analysis/Analytic/Basic.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Basic.lean:1477]: Is there a way to lint for this?

Comment (semorrison) [reply Mathlib/Analysis/Analytic/Basic.lean]: I think it would be tricky to lint for. There are valid tactics to run when you have multiple goals open (`rotate_right`, `solve_by_elim*`, ...).

File: Mathlib/Analysis/Analytic/Basic.lean
Comment (AntoineChambert-Loir) [inline Mathlib/Analysis/Analytic/Basic.lean:912]: For the docstring, I'd say "imply continuous" rather than "means".

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (AntoineChambert-Loir) [inline Mathlib/Analysis/Analytic/Prod.lean:27]: Is it possible to treat a bilinear/multilinear map rather than product?

File: Mathlib/Analysis/Analytic/Basic.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Basic.lean:912]: Done.

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (eric-wieser) [reply Mathlib/Analysis/Analytic/Prod.lean:27]: I don't understand the request, `prod.mk` isn't bilinear!

Comment (girving) [reply Mathlib/Analysis/Analytic/Prod.lean:27]: I'm not sure I follow the suggestion, but it seems like this would be best handled by a separate lemma that bilinear maps are analytic composed with this one, and then the chained version.  I'll be doing that in particular to handle the case of multiplication in the next PR.  But it seems like we want this case regardless (and certainly I use it directly all the time downstream).

Comment (AntoineChambert-Loir) [reply Mathlib/Analysis/Analytic/Prod.lean:27]: Indeed. I confused `prod` with product.

Comment (girving) [reply Mathlib/Analysis/Analytic/Prod.lean:27]: Out of scope for this PR, but coming... https://github.com/girving/ray/blob/01188729d12baffe1bc0d88340223ea6e77ce280/Ray/Analytic/Analytic.lean#L180

Comment (AntoineChambert-Loir) [reply Mathlib/Analysis/Analytic/Prod.lean:27]: Another product is possible, which takes two analytic functions E -> F and G -> H and considers the map E x G -> F x H.
Your product is a restriction to the diagonal.
Also, what about things like (Fin n -> E -> F) and its dependent variant (Fin n -> E n -> F n)?

Comment (ADedecker) [reply Mathlib/Analysis/Analytic/Prod.lean:27]: The usual way we organize things (e.g for `Continuous`) is to first prove that maps to a product are continuous iff each component is. Then it's easy to deduce that both notions of product work well. AFAIK there's no reason that it wouldn't work for analyticity (see `ContinuousMultilinearMap.prod`)

And indeed the same scheme works for both `Prod` and (finite? I'm not even sure we need it) `Pi` (see `ContinuousMultilinearMap.pi`).

Comment (ADedecker) [reply Mathlib/Analysis/Analytic/Prod.lean:27]: Regarding naming I think this is right though. We use `prod` to refer to the "diagonal" product (aka applying the universal property) and `prod_map` to refer to the one Antoine describes.

File: Mathlib/Analysis/Analytic/Basic.lean
Comment (sgouezel) [inline Mathlib/Analysis/Analytic/Basic.lean:1470]: could you only require `h` for elements of `N`?

File: Mathlib/Analysis/Analytic/Basic.lean
Comment (sgouezel) [inline Mathlib/Analysis/Analytic/Basic.lean:1471]:
```suggestion
    AnalyticAt 𝕜 (fun z ↦ ∑ n in N, f n z) c := by
```
is easier to read. (You may need to `open scoped BigOperators`)

File: Mathlib/Analysis/Analytic/Basic.lean
Comment (sgouezel) [inline Mathlib/Analysis/Analytic/Basic.lean:1478]: ditto.

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (sgouezel) [inline Mathlib/Analysis/Analytic/Prod.lean:34]: Normally, we use a single tactic per line (except in one line proofs). Here, you could use 
```
    refine ENNReal.le_of_forall_nnreal_lt (fun r rr ↦ ?_)
```
to get the same effect as your two tactics.

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (sgouezel) [inline Mathlib/Analysis/Analytic/Prod.lean:51]:
```suggestion
        intro y yr
        apply HasSum.prod_mk
```

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (sgouezel) [inline Mathlib/Analysis/Analytic/Prod.lean:53]: At this point of your proof, you have two goals, so you should focus on the first one and then the second one, with
```suggestion
        · exact fp.hasSum (EMetric.ball_subset_ball (min_le_left _ _) yr)
        · exact gq.hasSum (EMetric.ball_subset_ball (min_le_right _ _) yr) }
```

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (sgouezel) [inline Mathlib/Analysis/Analytic/Prod.lean:61]: The canonical name per our naming conventions would be
```suggestion
theorem AnalyticAt.comp₂ {h : F → G → H} {f : E → F} {g : E → G} {x : E}
```
Same thing below.

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (sgouezel) [inline Mathlib/Analysis/Analytic/Prod.lean:66]:
```suggestion
  rw [e]
```

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (sgouezel) [inline Mathlib/Analysis/Analytic/Prod.lean:77]:
```suggestion
    (fa : AnalyticAt 𝕜 (uncurry f) p) :
```
With my formulation, the lemma will apply in all situations (including the case of `(x, y)`), while your version will only apply when the point is already written as a product.

Same thing below.

**Review by sgouezel — COMMENTED** [review]
This looks very good, although I have many minor comments as expected. Thanks!

**Comment by ADedecker** [comment]
Note that there is a big overlap with #7552, which also proves analyticity of the product (cc @loefflerd).

**Comment by loefflerd** [comment]
> Note that there is a big overlap with #7552, which also proves analyticity of the product (cc @loefflerd).

Yes, indeed. @girving : how much of my PR is contained in this one of yours + others that will be following in the near future? If the answer is "all of it" then I'm happy to just delete my PR.

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Prod.lean:61]: Hmm that does seem like it's `defeq` to the version with explicit curry, but I should check to see if things infer correctly when doing `apply` in tactics.  +1 to it being way better if things do work.

(Thank you for this and the other comments!  Will check this and resolve the other changes later, but might be a couple day delay.)

**Comment by girving** [comment]
I see, that PR stomps on most/all of my mine.  In a good way, since they are the same, but: possibly I should just hold off submitting things for a while, unless something is interesting in taking a look through the code and figuring out if there is a useful thing to upstream that won't be immediately wasted effort on my part as I'm learning the system.  This is pure hobby stuff for me, so I often won't be able to turn things around rapidly.

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Prod.lean:61]: I checked, and this works great.  Fixed.

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Prod.lean:51]: Obsolete; `AnalyticAt.prod` went in in a separate PR.

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Prod.lean:53]: Obsolete.

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Prod.lean:77]: Fixed, thank you!

**Comment by girving** [comment]
Okay, I've rebased.  Now the results added are just the very simple ones, since `.prod` is gone.  Back to review!  Some questions I have:

1. What's the right name of `.along_fst` and `.along_snd`?  Is there a standard name to use?
2. Should I move things in `Prod.lean` to `Basic.lean` or `Linear.lean`, especially since `.prod` is now in `Basic`?

File: Mathlib/Analysis/Analytic/Basic.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Basic.lean:1582]: Oops, missed these comments.  Fixed, about to push.

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Prod.lean]: Will do in future (though apologies in advance as I expect to accidentally let a variety of them through as I do this a lot in the current code).  In any case, obsolete here as this function was already upstreamed by someone else.

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (loefflerd) [inline Mathlib/Analysis/Analytic/Prod.lean:25]: Where are the curried functions here? In this statement `h` is not curried. The curried form of `h` would be the map from `F` to functions `G → H` sending `x \in H` to the function `y ↦ h x y`. 

In nice situations (e.g. continuous functions on compacts) the continuity of `h` as a map on `F \times G` is equivalent to continuity of its curried form as a map `F \to C(G, H)`, so one can reduce continuity problems to a single variable (at a cost of allowing the target space to be large). But I'm not sure that's even defined here; the space of functions that are analytic at a given point (and arbitrary elsewhere) isn't a Banach space in any natural way, so I don't think we can define analyticity with this space as target.

Comment (loefflerd) [reply Mathlib/Analysis/Analytic/Prod.lean:25]: (Just to clarify, I have nothing against the content of these lemmas, I'm just advocating for a change to the docstrings.)

Comment (girving) [reply Mathlib/Analysis/Analytic/Prod.lean:25]: Sorry, will fix: they used to be curried and called `curry_comp`, but I fixed names and definitions to not be curried but not the doctoring.

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Prod.lean:1]: Should I call it `Constructions.lean` then, or something else?

Comment (loefflerd) [reply Mathlib/Analysis/Analytic/Prod.lean:1]: Yes, that sounds like a good name for it.

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Prod.lean:25]: I’m actually going to need that Banach space (the Linf version) later on to extend Hartogs theorem from 2D to ND, but I’ll build that when I use it.

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (alreadydone) [reply Mathlib/Analysis/Analytic/Prod.lean]:
```suggestion
    (fa : AnalyticAt 𝕜 (uncurry f) p) :
```
Side comment: several lemmas around [nhds_prod_eq](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Constructions.html#nhds_prod_eq) should probably also be "generalized" this way.

File: Mathlib/Analysis/Analytic/Prod.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Prod.lean:1]: Done, and while I was at it I added a few more lemmas about `inv` and `div` (and `Finset.prod`).

**Comment by digama0** [comment]
Note: I have pushed an update to the lean toolchain because this PR was on a [buggy version](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Potential.20data.20loss.20from.20.60lake.20clean.60.20with.204.2E2.2E0-rc2.2F3/near/397875701) of the toolchain. **WARNING**: checking out old commits of this PR using v4.2.0-rc2 or v4.2.0-rc3 can cause `lake clean` to delete your mathlib folder! If you need to do so, make sure to delete `lakefile.olean` manually before running any `lake` commands.

**Review by loefflerd — APPROVED** [review]
I pushed a small change which just gets rid of some repetition in the statements. Otherwise LGTM (but I am not a maintainer, so my word is not enough to merge).

**Comment by girving** [comment]
@loefflerd: Thank you!  Those changes look good.

File: Mathlib/Analysis/Analytic/Constructions.lean
Comment (sgouezel) [inline Mathlib/Analysis/Analytic/Constructions.lean:293]:
```suggestion
    (N : Finset α) (h : ∀ n ∈ N, AnalyticAt 𝕜 (f n) c) :
```
It's exactly the same thing, just more readable. Same thing below.

**Review by sgouezel — COMMENTED** [review]
Looks great, thanks!

File: Mathlib/Analysis/Analytic/Constructions.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Constructions.lean:167]: Done.

File: Mathlib/Analysis/Analytic/Constructions.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Constructions.lean:173]: Done.

File: Mathlib/Analysis/Analytic/Constructions.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Constructions.lean:195]: Done.

File: Mathlib/Analysis/Analytic/Constructions.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Constructions.lean:251]: Done.

File: Mathlib/Analysis/Analytic/Constructions.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Constructions.lean:293]: Done.

File: Mathlib/Analysis/Analytic/Linear.lean
Comment (girving) [reply Mathlib/Analysis/Analytic/Linear.lean:25]: Gone.

**Comment by sgouezel** [comment]
bors r+
Thanks!

**Pull request closed without merging**
